util/flocks/src/frwlock/mod.rs

```diff
@@ -0,0 +1,200 @@
+pub mod read_guard;
+pub mod write_guard;
+
+pub use write_guard::FrwLockWriteGuard;
+pub use read_guard::FrwLockReadGuard;
+use rustix::{
+    fs::{flock, FlockOperation},
+    fd::AsFd,
+};
+use std::cell::UnsafeCell;
+use thiserror::Error;
+use tokio::task::yield_now;
+
+#[derive(Debug, Error)]
+pub enum FrwLockError {
+    #[error("Lock is not available")]
+    LockNotAvailable,
+    #[error("File error: {0}")]
+    FileError(#[from] std::io::Error),
+}
+
+/// A file-based read-write lock.
+/// This only mutually excludes processes trying to violate the lock, not the same process--which is not considered contention.
+/// If you want to prevent contention within the same process, you should wrap this in your preferred synchronization primitive.
+pub struct FrwLock<T: AsFd> {
+    cell: UnsafeCell<T>,
+}
+
+impl<T: AsFd> FrwLock<T> {
+    pub fn new(file: T) -> Self {
+        Self {
+            cell: UnsafeCell::new(file),
+        }
+    }
+
+    pub(crate) fn try_write(&self) -> Result<FrwLockWriteGuard<T>, FrwLockError> {
+        let file = unsafe { &*self.cell.get() };
```

Comment:
The raw pointer obtained from `self.cell.get()` must be valid. Maybe this is the only way, but it seems unnecessary to me to be using `unsafe` blocks for a config. 

I didn't realise that with this design we would need to actually read and write to files themselves. Just some state in memory (the dynamic KV part). Would this be possible?

Reply:
You still need unsafe blocks for `UnsafeCell`. 

We need to persist, so that other processes can access and update.